### ISSUE TITLE
How to pass parameters to a middleware and influence its processing logic ?

### ISSUE THREAD

**Issue by GoodManWEN** [tiangolo/fastapi#2933]
### Description

Hello everyone, we know that through `request.state` we can pass in some custom data to handlers from middleware's before-handle process and thus influence its behaviour, I'm currently wondering how can a handler influence the middleware's logic after it.

My specific business scenario is that I have a routed address (let's take `/api` for example) whose function is to calculate a dynamic result that will consume a relatively long time and return a relatively large json response. Operationally, an effective way to improve efficiency is to use some buffer (like redis) to cache its results, in this way, the calculation time can be saved everytime the cache hits.

Since my memory is limited, I want to store the gzip-compressed bytes in a buffer instead of raw json streams, this will greatly increase the amount of cache I could handle. Specifically, because of there's a large number of `numbers` in the response, a response content is usually about 20MB in size without gzip, while after compression it is about only 1MB. This means that with 1GB of memory, I can only cache 50 different responses without compression, while with compression I can cache 1000, which is a significant difference that can hardly be ignored.

Because of these requirements, I wanted to implement a full-featured gzip middleware, but there were several technical confusions. The first is that I want to control the middleware whether to compress or not, apparently, if the response is not hitting the cache but is generated dynamically then it should be compressed, but on the contrary, it should not be compressed again since it has already been compressed once.  The second question is, even if I can control the middleware not to compress, how do I replace its results with bytes that have already been compressed when there's no need to run the compress logic?

### Example

Since I yet don't know how to implement, forgive me that I can only provide some pseudo-code to illustrate my thinking. 

The following code describes a relatively complex response address that does not contain middleware:
```Python
from fastapi import FastAPI, Request
import uvicorn

app = FastAPI()

@app.post("/api")
async def root(request :Request, some_args: int):

    # Using the Fibonacci series to simulate a time-consuming 
    # computational operation
    def fib_recur(n):
  	if n <= 1: return n
  	return fib_recur(n-1) + fib_recur(n-2)
    
    # In addition to the time-consuming calculation, the size of 
    # the returned content is also large.
    return {"response_content":  [fib_recur(31)] * 10000000 }

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8080)
```

I want to achieve the following after adding middleware:
```Python
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from utils import fib_recur, search_for_if_cache_hit
import uvicorn

app = FastAPI()

@app.post("/api")
async def root(request :Request, some_args: int):
    cache_hit_flag , cache_content = search_for_if_cache_hit('/api' , some_args)
    if cache_hit_flag == True:
        # Skip Calculation
        response = SomeKindOfResponse(message_content = cache_content)
        response.store['need_gzip'] = False
    else:
        # Calculate normally
        response = JSONResponse( {"response_content":  [fib_recur(31)] * 10000000 } )
        response.store['need_gzip'] = True
    return response

@app.middleware("http")
async def gzip_middleware(request: Request, call_next):
    response = await call_next(request)
    if response.store['need_gzip'] == True:
        response = gzip_handler(response)
    else:
        pass
    return response

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8080)
```

Thanks!

### Environment

* OS: [e.g. Linux / Windows / macOS]: Python3.8.6
* FastAPI Version [e.g. 0.3.0]: 0.63.0

**Comment by valentin994** [comment]
Have you taken [background tasks](https://fastapi.tiangolo.com/tutorial/background-tasks/) into consideration? It looks applicable for your use case here.

**Comment by GoodManWEN** [comment]
@valentin994 Thanks for the reply, but I'm senserly consider that this isn't the answer I need. Since I need to modify the return value for each request at the moment, but not do something in the background afterwards.

**Comment by juntatalor** [comment]
@GoodManWEN Hi!
I think that it is possible to pass params via request.state, because it is the same object in middleware and handler.
Like this:

```
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, Response
import random
import uvicorn

app = FastAPI()

def fib_recur(n):
  	if n <= 1: return n
  	return fib_recur(n-1) + fib_recur(n-2)


def search_for_if_cache_hit(p1, p2):
    if random.choice((0, 1)) == 1:
        return True, b'123'
    return False, b'321'

def gzip_handler(response: Response):
    # Gzip response.content here and modify response.headers
    return response


@app.get("/api")
async def root(request :Request, some_args: int):
    cache_hit_flag , cache_content = search_for_if_cache_hit('/api' , some_args)
    if cache_hit_flag == True:
        # Skip Calculation
        response = Response(cache_content)
        request.state.need_gzip = False
    else:
        # Calculate normally
        response = JSONResponse( {"response_content":  [fib_recur(3)] * 100 } )
        request.state.need_gzip = True
    return response


@app.middleware("http")
async def gzip_middleware(request: Request, call_next):
    response = await call_next(request)
    if getattr(request.state, 'need_gzip', None) == True:
        response = gzip_handler(response)
    return response

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8080)
```

**Comment by GoodManWEN** [comment]
@juntatalor Thanks! You reminded me that python dictionaries are structures that store pointers, which I had missed before. This means that indeed we can pass data to the middleware without relying on the return value of call_next. Now the only remaining question is how to get the middleware to return bytes directly without compression.

**Comment by juntatalor** [comment]
> Now the only remaining question is how to get the gzip middleware to return bytes directly without compression.

But isn't it already done by `need_gzip` flag? With false value the response is returned with no changes (raw json in example). The gzip operation should be performed in `gzip_handler` function.

**Comment by GoodManWEN** [comment]
@juntatalor  Yes indeed, however I'm not familiar with http's gzip protocal, nor many python web frameworks for python implemented gzip can provide a reference. I looked a little deeper into starlette's gzip plugin source code and it seems to involve certain ASGI implementation that are generally hidden and it seems that which cannot be modified directly at the request handler level.

**Comment by juntatalor** [comment]
I admit, this is more complicated than it seemed at first.
The first thing is that starlette's GzipMiddleware works with Messages, which are generated by Response class itself. So gzip middleware does not receive Response object at all. Also there is no information about Request state in messages:
```
# Response.__call__
await send(
            {
                "type": "http.response.start",
                "status": self.status_code,
                "headers": self.raw_headers,
            }
        )
        await send({"type": "http.response.body", "body": self.body})
```

I think it is simple to use contextvars (python 3.7+) to pass use_gzip flag into middleware. It is guaranteed to be local on current async stack. Here is a minimal working example, but code should be cleared a bit: 

```
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse, Response
import random
import uvicorn
from starlette.middleware.gzip import GZipMiddleware, GZipResponder
from starlette.datastructures import Headers

from contextvars import ContextVar

use_gzip = ContextVar('use_gzip')

app = FastAPI()

def fib_recur(n):
  	if n <= 1: return n
  	return fib_recur(n-1) + fib_recur(n-2)


def search_for_if_cache_hit(p1, p2):
    if random.choice((0, 1)) == 1:
        return True, b'123'
    return False, b'321'


@app.get("/api")
async def root(request :Request, some_args: int):
    cache_hit_flag , cache_content = search_for_if_cache_hit('/api' , some_args)
    if cache_hit_flag == True:
        # Skip Calculation
        response = Response(cache_content)
        use_gzip.set(False)
    else:
        # Calculate normally
        response = JSONResponse( {"response_content":  [fib_recur(3)] * 100 } )
        use_gzip.set(True)
    return response


class MyGZipResponder(GZipResponder):
    async def send_with_gzip(self, message) -> None:
        if use_gzip.get():
            await super(MyGZipResponder, self).send_with_gzip(message)
        else:
            await self.send(message)


class MyGzipMiddleware(GZipMiddleware):
    # subclass to modify GzipResponder -> MyGzipResponder
    async def __call__(self, scope, receive, send) -> None:
        if scope["type"] == "http":
            headers = Headers(scope=scope)
            if "gzip" in headers.get("Accept-Encoding", ""):
                responder = MyGZipResponder(self.app, self.minimum_size)
                await responder(scope, receive, send)
                return
        await self.app(scope, receive, send)

app.add_middleware(MyGzipMiddleware)

if __name__ == '__main__':
    uvicorn.run(app, host='0.0.0.0', port=8080)
```

**Comment by GoodManWEN** [comment]
@juntatalor Thanks for kind reply. I traced your code and I think we so far solved the problem. As a summary, the first problem we encountered was passing values to the middleware's post-handler logic. In the most basic way, we can called directly in the high-level api to make `request.state` transfer cross-sate pointers. This is a common solution, but this approach does not work since gzip a middleware based on the asgi.

As an answer to the question in the previous discussion, according to my understanding, we modify the logic of gzip by inheritance it and replace its `send_with_gzip` method, directly replace the message body with compressed result bytes could achieve the effect. 

So the only problem left is how to pass the data. `contextvars` is a very nice feature to solve this problem, which I didn't know about before, thanks for introducing it to me, and thanks for the dynamic ability of python.

I'm going to close this issue, thanks once again. 

btw I think you may not be aware that markdown allows you to highlight code by markup the language of code blocks, which significantly enhances the readability of the code, forgive me if you already know.

**Comment by tiangolo** [comment]
Thanks for the help here @valentin994 and @juntatalor ! 👏  🙇 

Thanks for reporting back and closing the issue @GoodManWEN  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
